current/runtime/storage.md

```diff
@@ -4,22 +4,128 @@ lang: en
+* [Storage Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html) - A key-value hash map
+* [Storage Double Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html) - An
+  implementation of a map with two keys that provides the important ability to efficiently remove all entries that have
+  a common first key
+* Iterable Storage Maps - Implementations of
+  [Storage Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageMap.html) and
+  [Storage Double Map](https://substrate.dev/rustdocs/master/frame_support/storage/trait.IterableStorageDoubleMap.html)
+  whose keys and values can be iterated over
 
-Any value which can be encoded by the [Parity SCALE codec](../advanced/codec) is supported by these storage APIs.
+The type of Storage Item you select should depend on the logical way in which the value will be used by your runtime.
 
-### Storage Declaration
+### Storage Value
 
-You can use the `decl_storage!` macro to easily create new runtime storage items. Here is an example of what it looks like to declare each type of storage item:
+This type of storage item should be used for values that are viewed as a single unit by the runtime, whether that is a
+single primitive value, a single `struct` or a single collection of related items. Although wrapping related items in
+a shared `struct` is an excellent way to reduce the number of storage reads (an important consideration), at some point
+the size of the object will begin to incur costs that may outweigh the optimization in storage reads. Storage values
+can be used to store lists of items, but runtime developers should take care with respect to the size of these lists.
+Large lists incur storage costs just like large `structs`. Furthermore, iterating over a large list in your runtime may
+result in exceeding the block production time.
+
+#### Methods
+
+Refer to the Storage Value documentation for
+[a comprehensive list of the methods that Storage Values expose](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#required-methods).
+Some of the most important methods are summarized here:
+
+##### `get()`
+
+Load the value from storage.
+
+##### `put(val)`
+
+Store the provided value.
+
+##### `mutate(fn)`
+
+Mutate the value with the provided function.
+
+##### `take()`
+
+Remove the value from storage.
+
+### Storage Maps
+
+Storage Maps are implemented as hash maps, which is a pattern that should be familiar to most developers. In order to
+give blockchain engineers increased control over the way in which these data structures are stored, Substrate allows
+developers to select the hashing algorithm that is used to generate map keys. Map data structures are ideal for managing
+sets of items whose elements will be accessed randomly, as opposed to iterating over them sequentially in their entirety.
```

Comment:
I think it would be better to show how it is implemented in the underlying trie to make it more understandable about advantage/disadvantage.

Basically the value is put in the trie at the path==some_storage_prefix++hashed_key.
So each value is stored in an individual node, so reading/writing a value only decode/encode one raw_value which is stored in one trie-node.

also the performance of the map will depends on how big the map is because read/writing one value will recompute the path==some_prefix++hashed_key. and the hashed_key part can be more or less costy to compute. https://www.shawntabrizi.com/substrate/transparent-keys-in-substrate/

Reply:
There is [a separate document that goes over how storage is implemented](https://github.com/substrate-developer-hub/knowledgebase/blob/runtime-storage/current/advanced/storage.md). This document is meant to explain Substrate's storage interfaces and help blockchain runtime developers understand when and how to use them.

Reply:
this document will not explain how frame_support::storage abstraction are implemented in storage do they ? so I would explain their implementation here actually.

Reply:
I believe the purpose of the advance document is to cover implementation and the purpose of this document (the runtime document) is to cover the interface. @joepetrowski can you provide your input?